source/MRMesh/MRIntersection.h

```diff
@@ -1,21 +1,141 @@
 #pragma once
 
+#include "MRMeshFwd.h"
 #include "MRPlane3.h"
 #include "MRLine3.h"
+#include "MRLineSegm.h"
 #include <optional>
 
 namespace MR
 {
 
+template<typename T>
+const T errorLimit = std::numeric_limits<T>::epsilon() * 20.f;
+
+// finds an intersection between a plane1 and a plane2
+// returns nullopt if they are parallel
+// returns line from plane1 if plane1 equal plane2
+template<typename T>
+std::optional<Line3<T>> intersection( const Plane3<T>& plane1, const Plane3<T>& plane2 )
+{
+    const auto plane1N = plane1.normalized();
+    const auto plane2N = plane2.normalized();
+    const auto crossDir = cross( plane1N.n, plane2N.n );
+
+    const auto planeNpoint = plane1N.n * plane1N.d;
+    const auto plane2Npoint = plane2N.n * plane2N.d;
+
+    if ( crossDir.length() < errorLimit<T> )
+    {
+        if ( ( planeNpoint - plane2Npoint ).length() < errorLimit<T> )
+        {
+            auto bv = plane1N.n.furthestBasisVector();
+            return Line3<T>( planeNpoint, cross( plane1N.n, bv ).normalized() );
+        }
+        else
+            return {};
+    }
+
+    const auto p1 = plane1N.project( plane2Npoint );
+    const auto p2 = plane2N.project( planeNpoint );
+    const auto p3 = Line3<T>(plane2Npoint, ( p1 - plane2Npoint ) ).project(p2);
+    return Line3<T>( plane2Npoint + ( p2 - plane2Npoint ) * ( p1 - plane2Npoint ).length() / ( p3 - plane2Npoint ).length(), crossDir.normalized() );
+}
+
 // finds an intersection between a plane and a line;
 // returns nullopt if they are parallel
 template<typename T>
 std::optional<Vector3<T>> intersection( const Plane3<T> & plane, const Line3<T> & line )
 {
-    const auto den = dot( plane.n, line.d );
+    const auto planeN = plane.normalized();
+    const auto lineN = line.normalized();
```

Comment:
Was the function incorrect without this?

Reply:
removed. added comments